cartridge/confapplier.lua

```diff
@@ -16,362 +20,460 @@ local pool = require('cartridge.pool')
+    ['BoxConfigured'] = {'ConnectingFullmesh'},
+
+-- normal operation
+    ['ConnectingFullmesh'] = {'ConfiguringRoles', 'OperationError'},
+    ['ConfiguringRoles'] = {'RolesConfigured', 'OperationError'},
+    ['RolesConfigured'] = {'ConnectingFullmesh'},
+
+-- errors
+    ['InitError'] = {},
+    ['BootError'] = {},
+    ['OperationError'] = {}, -- {'BoxConfigured'}
+    -- Disabled
+    -- Expelled
+}
+local function set_state(new_state, err)
+    checks('string', '?')
+    StateError:assert(
+        utils.table_find(_transitions[vars.state], new_state),
+        'invalid transition %s -> %s', vars.state, new_state
+    )
+
+    if new_state == 'InitError'
+    or new_state == 'BootError'
+    or new_state == 'OperationError'
+    then
+        log.error('Instance entering failed state: %s -> %s\n%s',
+            vars.state, new_state, err
+        )
+    else
+        log.info('Instance state changed: %s -> %s',
+            vars.state, new_state
+        )
+    end
 
+    vars.state = new_state
+    vars.error = err
+end
 
---- Load configuration from the filesystem.
--- Configuration is a YAML file.
--- @function load_from_file
+--- Validate configuration by all roles.
+-- @function validate_config
 -- @local
--- @tparam ?string filename Filename to load.
--- When omitted, the active configuration is loaded from `<workdir>/config.yml`.
--- @treturn[1] table
+-- @tparam table clusterwide_config_new
+-- @treturn[1] boolean true
 -- @treturn[2] nil
 -- @treturn[2] table Error description
-local function load_from_file(filename)
-    checks('?string')
-    filename = filename or fio.pathjoin(vars.workdir, 'config.yml')
-
-    if not utils.file_exists(filename) then
-        return nil, e_config_load:new('file %q does not exist', filename)
+local function validate_config(clusterwide_config, _)
+    checks('ClusterwideConfig', 'nil')
+    assert(clusterwide_config.locked)
+
+    local conf_new = clusterwide_config:get_readonly()
+    local conf_old
+    if vars.clusterwide_config then
+        conf_old = vars.clusterwide_config:get_readonly()
+    end
+    if conf_old == nil then
+        conf_old = {}
     end
 
-    local raw, err = utils.file_read(filename)
-    if not raw then
+    local ok, err = ddl_manager.validate_config(conf_new, conf_old)
+    if not ok then
         return nil, err
     end
 
-    local confdir = fio.dirname(filename)
+    return roles.validate_config(conf_new, conf_old)
+end
 
-    local conf, err = e_yaml:pcall(yaml.decode, raw)
-    if not conf then
-        if not err then
-            return nil, e_config_load:new('file %q is empty', filename)
-        end
 
-        return nil, err
-    end
+--- Apply the role configuration.
+-- @function apply_config
+-- @local
+-- @tparam table conf
+-- @treturn[1] boolean true
+-- @treturn[2] nil
+-- @treturn[2] table Error description
+local function apply_config(clusterwide_config)
+    checks('ClusterwideConfig')
+    assert(clusterwide_config.locked)
+    assert(
```

Comment:
> Why do you check it here?

To be sure I didn't make a mistake *before* calling `box.cfg`. It's just an assertion and I hope it will never raise.